platform-sdk/docs/core/tipset-algorithm.md

```diff
@@ -0,0 +1,260 @@
+
+Find the partial weighted advancement score between X and Y with a target of node A.
+
+X = [1, 3, 5, 2]
+Y = [7, 3, 2, 11]
+
+The entry for node A advances, but we ignore it because A is the target.
+The entry for D advances, add 2. Total weighed advancement score is 2.
+```
+
+# Snapshot Improvement Scores
+
+Each new event is assigned a tipset improvement score. Each node assigns the events that it creates an
+improvement score. Nodes do not assign events created by other nodes an improvement score.
+
+Keep track of a special tipset, called the "snapshot" tipset. The snapshot starts out empty, i.e.
+`[-1, -1, -1, ..., -1]` at genesis. Periodically, the snapshot is updated to a more recent tipset
+(how this happens is described below).
+
+Each time a node creates a new event, compare that event's tipset to the snapshot tipset, and find the partial
+weighted advancement score of that new event targeting the event's creator.
+
+There is a special threshold that causes the snapshot to advance. The threshold for each node is slightly different,
+unless nodes have the same consensus weight. To calculate the threshold, find the minimum weight required to have >2/3
+of the total weight, and then subtract the node's weight.
+
+```
+Example of computing snapshot advancement threshold
+
+Suppose an address book has the following weights: A = 5, B = 9, C = 11, D = 2. Compute the snapshot advancement
+threshold for node A.
+
+Total consensus weight is 27. The amount of stake required to have >2/3 of the total weight is 19. Subtract A's weight,
+and A's snapshot advancement threshold is 14.
+```
+
+When a new event has a snapshot improvement score that is greater than or equal to the snapshot advancement threshold,
+then the snapshot is updated to the tipset of that new event.
+
+# Limiting Event Creation by Snapshot Improvement Score
+
+Nodes are not permitted to create a new event unless one of the following two conditions are true:
+
+1) this is the first event a node has ever created (i.e. this is genesis), or
+2) the snapshot improvement score exceeds the snapshot improvement score of the previous event, or
+3) the snapshot improvement score is greater than zero and this is the first event created since
+   we last updated the snapshot
+
+By following this rule, this ensures that each node stops creating new events in a finite amount of time if that
+node no longer is in communicating with a quorum of its peers.
+
+```
+Example
+
+Address book weights: A = 5, B = 9, C = 11, D = 2
+Consider from A's perspective. A's snapshot advancement threshold is 14.
+
+Initially, the snapshot is [-1, -1, -1, -1].
+
+A creates a genesis event with tipset [0, -1, -1, -1]. The snapshot improvement score is 0, which is ok since
+this is a genesis event. (It will never again be ok to have an advancement score of 0.)
+
+Next, A creates an event with self parent [0, -1, -1, -1] and other parent [-1, 2, -1, -1] with generation 2.
```

Comment:
Consider code ticks around node ID letter when written in sentences e.g.

"Next, `A` creates an event..."

Reply:
I've done this for all letter variables not in the example code blocks. These blocks use don't allow internal formatting since they already use the triple tics.